extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,167 @@
+import {RTC_VENDORS} from './callout-vendors.js';
+import {tryParseJson} from '../../../src/json';
+import {dev} from '../../../src/log';
+import {Services} from '../../../src/services';
+import {isArray, isObject} from '../../../src/types';
+import {isSecureUrl, parseUrl} from '../../../src/url';
+
+/** Timeout in millis */
+const DEFAULT_RTC_TIMEOUT = 1000;
+
+/** @type {string} */
+const TAG = 'real-time-config';
+
+/** @type {number} */
+const MAX_RTC_CALLOUTS = 5;
+
+export class RealTimeConfigManager {
+  constructor(element, win, ampDoc) {
+    this.element = element;
+    this.win = win;
+    this.rtcConfig = null;
+    this.calloutUrls = [];
+    this.urlReplacements_ = Services.urlReplacementsForDoc(ampDoc);
+  }
+
+  executeRealTimeConfig(customMacros) {
+    if (!this.rtcConfig) {
+      return null;
+    }
+    const rtcTimeout = this.rtcConfig.timeoutMillis || DEFAULT_RTC_TIMEOUT;
+    this.inflatePublisherUrls(customMacros);
+    this.inflateVendorUrls();
+    if (this.calloutUrls.length == 0) {
+      return Promise.resolve();
+    }
+    let rtcTime;
+    return Promise.all(this.calloutUrls.map(url => {
+      const rtcStartTime = Date.now();
+      /** ISSUE: For very short timeouts, and/or for slow devices, the calculated time
+       * that RTC took (rtcTime below) can be longer than the timeout without getting timed-out,
+       * due to the time it takes to set up the timeout promise.
+       * EXAMPLE, let rtcTimeout = 100ms.
+       * Time = 0: const rtcStartTime = 0
+       * Time = 10: call Services.timerFor.timeoutPromise
+       * Time = 25: timeout promise starts the xhr callout, timeout promise sees start time as 25
+       * Time = 120: xhr callout returns, timeoutPromise sees total time as 95ms, so no timeout
+       * Time = 125: rtcTime = 125ms, which is greater than the publisher specified timeout.
+       */
+      return Services.timerFor(this.win).timeoutPromise(
+          rtcTimeout,
+          Services.xhrFor(this.win).fetchJson(
+              url, {credentials: 'include'}).then(res => {
+                // Non-200 status codes are forbidden for RTC.
+                // TODO: Add to fetchResponse the ability to
+                // check for redirects as well.
+                if (res.status != 200) {
```

Comment:
Redirects are already not allowed by xhr service as it checks response.ok which is defined as response code 200-299 (https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) and thus does not include 302.  In the event you get a non-200, it will throw so I'm not sure this is necessary